Apps/UnitTests/Shared/Shared.cpp

```diff
@@ -99,39 +99,77 @@ TEST(JavaScript, All)
 }
 
 /*
-This test does a serie of initialization and shutdowns.
-It needs the shutdown PR to be merged before running properly.
-TEST(NativeAPI, LifeCycle)
+Repro for shutdown scenario
+Engine is shutdown before shader is finished compiling
+TEST(Shutdown, AsyncShaderCompilation)
 {
-    for (int cycle = 0; cycle < 20; cycle++)
-    {
-        Babylon::Graphics::Device device{deviceConfig};
-        std::optional<Babylon::Polyfills::Canvas> nativeCanvas;
+    std::promise<int32_t> exitCodePromise;
 
-        Babylon::AppRuntime runtime{};
-        runtime.Dispatch([&device, &nativeCanvas](Napi::Env env) {
-            device.AddToJavaScript(env);
+    Babylon::Graphics::Device device{ deviceConfig };
+    auto update{ device.GetUpdate("update") };
+    std::optional<Babylon::Polyfills::Canvas> nativeCanvas;
+
+    Babylon::AppRuntime runtime{};
+    runtime.Dispatch([&device, &nativeCanvas, &exitCodePromise](Napi::Env env) {
+        device.AddToJavaScript(env);
 
-            Babylon::Polyfills::XMLHttpRequest::Initialize(env);
-            Babylon::Polyfills::Console::Initialize(env, [](const char* message, auto) {
-                printf("%s", message);
-                fflush(stdout);
+        Babylon::Polyfills::XMLHttpRequest::Initialize(env);
+        Babylon::Polyfills::Console::Initialize(env, [](const char* message, auto) {
+            std::cout << message << std::endl;
+            std::cout.flush();
             });
-            Babylon::Polyfills::Window::Initialize(env);
-            nativeCanvas.emplace(Babylon::Polyfills::Canvas::Initialize(env));
-            Babylon::Plugins::NativeEngine::Initialize(env);
+        Babylon::Polyfills::Window::Initialize(env);
+
+        Babylon::Polyfills::XMLHttpRequest::Initialize(env);
+
+        nativeCanvas.emplace(Babylon::Polyfills::Canvas::Initialize(env));
+
+        Babylon::Plugins::NativeEngine::Initialize(env);
+
+        auto setExitCodeCallback = Napi::Function::New(
+            env, [&exitCodePromise](const Napi::CallbackInfo& info) {
+                Napi::Env env = info.Env();
+                exitCodePromise.set_value(info[0].As<Napi::Number>().Int32Value());
+            }, "setExitCode");
+        env.Global().Set("setExitCode", setExitCodeCallback);
         });
 
-        Babylon::ScriptLoader loader{runtime};
-        loader.LoadScript("app:///Scripts/babylon.max.js");
-        loader.LoadScript("app:///Scripts/babylonjs.materials.js");
+    Babylon::ScriptLoader loader{ runtime };
+    loader.LoadScript("app:///Scripts/babylon.max.js");
+    loader.LoadScript("app:///Scripts/babylonjs.materials.js");
+    loader.Eval(R"(
+            function CreateBoxAsync(scene) {
```

Comment:
Maybe add a comment explaining why this needs to be an async function rather than just synchronous setup code? Not clear to me.